text/2394-async_await.md

```diff
@@ -0,0 +1,582 @@
+
+在獲得基於 future 的生態系統的經驗和使用者的回饋後，我們發現了某些人因工程學挑戰。使用需要在等待點（await point）之間共享的狀態是非常不符合人因工程學的——需要 Arc 或 join chaining —— 雖然組合器通常比手動撰寫 future 更符合人因工程學，但它們仍然經常導致混亂的嵌套和 chained callbacks。
+
+幸運的是，Future 抽象非常適合與一種語法糖一起使用，該語法糖在許多具有非同步 IO 的語言中變得很常見 - async 和 await 關鍵字。簡而言之，非同步函式返回一個 future，而不是在呼叫它時立即執行。在函式內部，可以使用 await 表達式等待其他 future，這使它們在輪詢 future 時讓出控制權。從使用者的角度來看，他們可以像使用同步程式碼一樣使用 async/await，並且只需要注釋他們的函式和調用。
+
+Async/await 和 futures 通常是非同步和並行的強大抽象，並且可能在非同步 IO 空間之外有其他應用。我們今天遇到的案例通常與非同步 IO 相關，但透過引入一等公民語法和 libstd 的支援，我們相信更多的 async 和 await 案例也會蓬勃發展，而這些案例不直接與非同步 IO 相關。
+
+# 教學式解說
+[教學式解說]: #guide-level-explanation
+
+## 非同步函式
+
+函式可以添加 `async` 關鍵字，使它們成為「非同步函式」：
+
+```rust
+async fn function(argument: &str) -> usize {
+     // ...
+}
+```
+
+非同步函式的工作方式與普通函式不同。呼叫非同步函式時，它不會立即進入主體。相反，它執行實作 `Future` 特徵的匿名型別。當輪詢該 future 時，該函式被執行到它內部的下一個 `await` 或返回點（請參閱接下來 await 語法部分）。
+
+非同步函式是延遲計算的一種 - 在您開始輪詢函式返回的 future 之前，函式本體中沒有任何內容被實際執行。例如：
+
+```rust
+async fn print_async() {
+     println!("Hello from print_async")
+}
+
+fn main() {
+     let future = print_async();
+     println!("Hello from main");
+     futures::executor::block_on(future);
+}
+```
+
+這將在印出 `"Hello from main"` 之前印出 `"Hello from print_async"`。
+
+`async fn foo(args..) -> T` 是 `fn(args..) -> impl Future<Output = T>` 型別的函式。返回型別是編譯器產生的匿名型別。
+
+### `async ||` closures
+
+除了函式，非同步也可以應用於 closure。與非同步函式一樣，非同步 closure 的返回型別為 `impl Future<Output = T>`，而不是 `T`。當您呼叫該 closure 時，它會立即返回一個 future 而不執行任何內容（就像非同步函式一樣）。
+
+```rust
+fn main() {
+    let closure = async || {
+         println!("Hello from async closure.");
+    };
+    println!("Hello from main");
+    let future = closure();
+    println!("Hello from main again");
+    futures::block_on(future);
+}
+```
+
+這將在印出 `"Hello from async closure."` 之前印出兩個 `"Hello from main"`。
+
+`async` closure 可以用 `move` 來捕獲它們包覆在 closure 內的變數的所有權。
+
+## `async` 區塊
+
+您可以使用 `async` 區塊直接將 future 創建為表達式：
+
+```rust
+let my_future = async {
+    println!("Hello from an async block");
+};
+```
+
+這種形式幾乎等同於立即呼叫的 `async` closure。即是：
+
+```rust
+async { /* body */ }
+
+// is equivalent to
+
+(async || { /* body */ })()
+```
+
+除了像 `return`、`break` 和 `continue` 這樣的控制流結構不允許在 `body` 中使用（除非它們出現在一個新的控制流上下文中，比如 closure 或 loop）。 `?`-operator 和 early return 如何在非同步區塊中工作尚未確定（請參閱未解決的問題）。
+
+與 `async` closure 一樣，`async` 區塊可以添加 `move`，以捕獲區塊內所包覆的變數的所有權。
+
+## 編譯器內嵌的 `await!`
+
+一個名為 `await!` 的內嵌函式被添加到編譯器。`await!` 可用於「暫停」future 的計算，將控制權交還給呼叫者。`await!` 接受任何實作 `IntoFuture` 的表達式，並計算為此 future 所傳入的泛型型別（如下面範例的 `Output`）之值。
+
+```rust
+// future: impl Future<Output = usize>
+let n = await!(future);
+```
+
+await 的擴展在它接收到的 future 上重複呼叫 `poll`，當它返回 `Poll::Pending` 時產生對函式的控制，並最終在它返回 `Poll::Ready` 時獲取該項目值。
+
+`await!` 只能在非同步函式、closure 或區塊內使用。在該上下文之外使用它是一個錯誤。
+
+（`await!` 是一個內嵌編譯器，為以後確定其確切語法留出空間。可在未解決的問題部分中查看更多訊息。）
```

Comment:
Original: (await! is a compiler built-in to leave space for deciding its exact syntax later. See more information in the unresolved questions section.)

```
（`await!` 是編譯器的內建方法，為以後確定其確切語法留出彈性空間。詳細資訊請參閱〈未解決的問題〉部分。）
```